File: csharp/need-for-speed/NeedForSpeed.cs
class RemoteControlCar
{
    internal int speed;
    internal readonly int batteryDrain;

    private int distance;
    internal int batteryStatus = 100;

    public RemoteControlCar(int speed, int batteryDrain)
    {
        this.speed = speed;
        this.batteryDrain = batteryDrain;
    }

    public bool BatteryDrained() =>
        batteryStatus < batteryDrain;

    public int DistanceDriven() => distance;

    public void Drive()
    {
        if (BatteryDrained()) {
            return;
        }

        batteryStatus -= batteryDrain;
        distance += speed;
    }

    public static RemoteControlCar Nitro() =>
        new RemoteControlCar(50, 4);
}

class RaceTrack
{
    private readonly int distance;

    public RaceTrack(int distance)
    {
        this.distance = distance;
    }

    public bool TryFinishTrack(RemoteControlCar car)
    {
        var runs = 100 / car.batteryDrain;
        var distance = runs * car.speed;

        return distance >= this.distance;
    }
}

File: java/need-for-speed/src/main/java/NeedForSpeed.java
class NeedForSpeed {
  private int battery = 100;
  private int distance = 0;

  private int speed;
  private int batteryDrain;

  NeedForSpeed(int speed, int batteryDrain) {
    this.speed = speed;
    this.batteryDrain = batteryDrain;
  }

  public boolean batteryDrained() {
    return battery <= 0;
  }

  public int distanceDriven() {
    return distance;
  }

  public void drive() {
    if (!batteryDrained()) {
      distance += speed;
      battery -= batteryDrain;
    }
  }

  public static NeedForSpeed nitro() {
    return new NeedForSpeed(50, 4);
  }
}

class RaceTrack {
  private int distance;

  RaceTrack(int distance) {
    this.distance = distance;
  }

  public boolean tryFinishTrack(NeedForSpeed car) {
    while (!car.batteryDrained()) {
      car.drive();
    }

    return car.distanceDriven() >= distance;
  }
}
